pkg/tfbridge/schema_test.go

```diff
@@ -1241,92 +1241,165 @@ func TestInvalidAsset(t *testing.T) {
-		"pulumi_override_tf_string_to_int":        {Type: schemav1.TypeString},
-		"pulumi_override_tf_string_to_integer":    {Type: schemav1.TypeString},
-		"tf_empty_string_to_pulumi_bool_override": {Type: schemav1.TypeString},
+	// We need to assert that when both the Pulumi type (String) and the Terraoform
+	// type (Int) are large enough to hold a large number, we never round trip it
+	// through a smaller type like a float64.
+	//
+	// We assert this requirement by checking that the number we use *does not* round
+	// trip through float64.
+	t.Run("number_is_large", func(t *testing.T) {
+		t.Parallel()
+		assert.NotEqual(t, largeNumber, int64(float64(largeNumber)))
 	})
 
-	typeOverrides := map[string]*SchemaInfo{
-		"pulumi_override_tf_string_to_boolean": {
-			Type: "boolean",
+	tests := []struct {
+		name string
+
+		tfSchema *schemav1.Schema
+		info     *SchemaInfo
+
+		tfInput  any
+		tfOutput resource.PropertyValue
+	}{
+		{
+			name: "pulumi_override_tf_string_to_boolean",
+
+			tfSchema: &schemav1.Schema{Type: schemav1.TypeString},
+			info:     &SchemaInfo{Type: "boolean"},
+
+			tfInput:  MyString("true"),
+			tfOutput: resource.NewProperty(true),
 		},
-		"pulumi_override_tf_string_to_bool": {
-			Type: "bool",
+		{
+			name: "pulumi_override_tf_string_to_bool",
+
+			tfSchema: &schemav1.Schema{Type: schemav1.TypeString},
+			info:     &SchemaInfo{Type: "bool"},
+
+			tfInput:  MyString("true"),
+			tfOutput: resource.NewProperty(true),
 		},
-		"pulumi_empty_tf_override": {
-			Type: "",
+		{
+			name: "pulumi_empty_tf_override",
+
+			tfSchema: &schemav1.Schema{Type: schemav1.TypeString},
+			info:     &SchemaInfo{Type: ""},
+
+			tfInput:  MyString("true"),
+			tfOutput: resource.NewProperty("true"),
 		},
-		"pulumi_override_tf_string_to_int": {
-			Type: "int",
+		{
+			name: "pulumi_override_tf_string_to_int",
+
+			tfSchema: &schemav1.Schema{Type: schemav1.TypeString},
+			info:     &SchemaInfo{Type: "int"},
+
+			tfInput:  MyString("1"),
+			tfOutput: resource.NewProperty(1.0),
 		},
-		"pulumi_override_tf_string_to_integer": {
-			Type: "integer",
+		{
+			name: "pulumi_override_tf_string_to_integer",
+
+			tfSchema: &schemav1.Schema{Type: schemav1.TypeString},
+			info:     &SchemaInfo{Type: "integer"},
+
+			tfInput:  MyString("1"),
+			tfOutput: resource.NewProperty(1.0),
 		},
-		"tf_empty_string_to_pulumi_bool_override": {
-			Type:           "boolean",
-			MarkAsOptional: boolPointer(true),
+		{
+			name: "tf_empty_string_to_pulumi_bool_override",
+
+			tfSchema: &schemav1.Schema{Type: schemav1.TypeString},
+			info: &SchemaInfo{
+				Type:           "boolean",
+				MarkAsOptional: True(),
+			},
+
+			tfInput:  MyString(""),
+			tfOutput: resource.NewNullProperty(),
+		},
+		{
+			name: "tf_int_to_pulumi_string",
+
+			tfSchema: &schemav1.Schema{Type: schemav1.TypeInt},
+			info:     &SchemaInfo{Type: "string"},
+
+			tfInput:  largeNumber,
+			tfOutput: resource.NewProperty(strconv.FormatInt(largeNumber, 10)),
```

Comment:
So when Terraform gives us a large number, we transform it into a string? But if the number is _small_, we transform to an int64, but one of a certain size?

I'm confused because both `largeNumber` as well as other tf.Output values all look like int64. I presume this is a property of shim.TypeInt?

Reply:
> So when Terraform gives us a large number, we transform it into a string? But if the number is small, we transform to an int64, but one of a certain size?

We never do anything based on the size of the number given. If the TF type is an Int (`int64` in Go) and the Pulumi type is `"string"` we ensure that the engine sees a string and that the upstream provider sees an `int64`. The test here is to make sure we do the conversion without going through a datatype that cannot express all of `int64` (like `float64`).

> I'm confused because both `largeNumber` as well as other tf.Output values all look like int64. I presume this is a property of shim.TypeInt?

`strconv.FormatInt` is of type `func(n int64, base int) string`, so `tfOutput` holds a string `resource.PropertyValue`.